### DISCUSSION TITLE
Implementation of 3rd party library breaks on page refresh

### DISCUSSION THREAD

**Discussion by Justincale** [oqtane/oqtane.framework#3458] [General]
Hi, I am developing a module that requires a modal so i have added the Blazor.Bootrap client library as i have used this previously with other blazor projects. I have followed information from various sources on how to install 3rd party component libraries (such as this post (https://github.com/oqtane/oqtane.framework/discussions/906#discussioncomment-137624) and have it running somewhat succesfully.

When the page containing the 3rd pary component first loads (after starting the project from Visual Studio for instance), the component works fine but fails with a jsInterop exception ("boostrap not defined") after I refresh the page. It looks to be some kind of timing issues with js references being loaded too late after a refresh?? 

Another reason I think it is related to timing is that if I place the control within a condition to only display when the pagestate is in edit mode, then I observe the following:

1 Refresh page whilst in standard pagestate view mode
2 Switch pagestate to edit mode: Control is loaded and works fine 
3 Refresh while pagestate is in edit mode: Control fails with jsinterop exception

I'm new to oqtane so was wondering if anyone might know what is happening here, as it's driving me nuts.

Thanks

JC

**Comment by sbwalker** [comment]
Oqtane already includes Bootstrap in its default themes - both the CSS and JS references for Bootstrap 5.3. So I am not sure if the Bootstrap.Client component library will try to register the Bootstrap CSS and JS again - perhaps with a different version? Bootstrap makes it very simple to create a modal with standard HTML markup: https://getbootstrap.com/docs/5.0/components/modal/

**Comment by artmedia** [comment]
No third party libraries are needed to open modal BS
This could be an example:

############ in Client project ##############
+++++++++++ add to Interop.cs ++++++++++

public async Task CloseModal(string idModale)
{
        await _jsRuntime.InvokeVoidAsync("Artmedia.AlbalaghaNews.CloseModal", idModale);
}

public async Task OpenModal(string idModale)
{
        await _jsRuntime.InvokeVoidAsync("Artmedia.AlbalaghaNews.OpenModal", idModale);
}


+++++++++++ add to index.razor +++++++++++ 
private async Task OpenModalEdit()
{
    var interop = new Artmedia.AlbalaghaNews.Interop(JSRuntime);
    await interop.OpenModal("modalEdit");
}


############ in Server project ############ 
+++++++++++ add to Module.js +++++++++++ 

Artmedia.AlbalaghaNews = {

    CloseModal: function (idModale) {
        let myModalObj = document.querySelector('#' + idModale);
        let myModalInstance = bootstrap.Modal.getOrCreateInstance(myModalObj); // Returns a Bootstrap modal instance
        myModalInstance.hide();
        console.log('CloseModal ' + '#' + idModale);
    },
    OpenModal: function (idModale) {
        let myModalObj = document.querySelector('#' + idModale);
        let myModalInstance = bootstrap.Modal.getOrCreateInstance(myModalObj); // Returns a Bootstrap modal instance
        myModalInstance.show();
        console.log('OpenModal ' + '#' + idModale);
    }
};
